File: 261-GraphValidTree/261-GraphValidTree.py
# Last updated: 4/29/2026, 10:32:05 AM
class Solution:
    def validTree(self, n: int, edges: List[List[int]]) -> bool:
        adj = {}
        for start, end in edges:
            if adj.get(start) == None:
                adj[start] = set()
            adj[start].add(end)
            if adj.get(end) == None:
                adj[end] = set()
            adj[end].add(start)
        visited = set()
        
        queue = [0]
        while len(queue) != 0:
            newQueue = []
            for start in queue:
                if start in visited:
                    return False
                visited.add(start)
                ends = adj.get(start)
                if ends != None:
                    for end in ends:
                        newQueue.append(end)
                        adj[end].remove(start)
            queue = newQueue
        return len(visited) == n
                

File: 261-GraphValidTree/261-GraphValidTree.js
// Last updated: 4/29/2026, 10:32:10 AM
/**
 * @param {number} n
 * @param {number[][]} edges
 * @return {boolean}
 */
// 2022-11-03
// X
// O(n)
// O(n)
var validTree = function(n, edges) {
    const adj = new Array(n).fill(null).map(() => new Set());
    const queue = [ 0 ];
    const visited = new Set();
    for(const [u, v] of edges) {
        adj[u].add(v);
        adj[v].add(u);
    }
    console.log(adj);
    
    while(queue.length !== 0) {
        const start = queue.shift();
        
        if(visited.has(start)) {
            return false;
        }
        visited.add(start);
        const ends = adj[start];
        if(ends) {
            for(const end of adj[start]) {
                queue.push(end);
                adj[end].delete(start);
            }    
        }
    }
    return visited.size === n;
}

var validTree = function(n, edges) {
    const adj = new Array(n).fill(null).map(() => new Set());
    const visited = new Set();
    for(const [u, v] of edges) {
        adj[u].add(v);
        adj[v].add(u);
    }
    console.log(adj);
    
    let queue = [ 0 ];
    while(queue.length !== 0) {
        const newQueue = [];
        for(const start of queue) {
            if(visited.has(start)) {
                return false;
            }
            visited.add(start);
            const ends = adj[start];
            if(ends) {
                for(const end of ends) {
                    newQueue.push(end);
                    adj[end].delete(start);
                }
            }
        }
        queue = newQueue;
    }
    return visited.size === n;
}